More batch mode driver work

graydon opened the pull request:
Followon to #14034, just more pieces of the batch-mode-driver puzzle. Still nothing wired up here yet, nor any changes to observable behaviour. Adds a change that fixes a typo in #14034 along the way.

graydon committed:
[BatchMode] Add -batch-mode driver option.

graydon committed:
[BatchMode] Add BatchModeCompile, multi-input form of StandardCompile.

graydon committed:
[BatchMode] Fix mis-chosen name in previous PR, NFC.

graydon committed:
[BatchMode] Add BatchJob subclass of Job, for use with BatchModeCompile.

graydon commented:
@swift-ci please test

jrose-apple commented on include/swift/Option/Options.td:
```diff
@@ -619,6 +619,10 @@ def whole_module_optimization : Flag<["-"], "whole-module-optimization">,
   Flags<[FrontendOption, NoInteractiveOption]>;
 
+def batch_mode : Flag<["-"], "batch-mode">,
+  Flags<[FrontendOption, NoInteractiveOption]>,
```
Please mark this HelpHidden for now. Also, *is* it a FrontendOption? Didn't @davidungar  already add one?

graydon replied:
@jrose-apple There is currently no other mention of batch mode in FrontendOptions.td or Options.td. There might be in one of the changesets that @davidungar has open.

I do not otherwise understand the question: do you want it to be added to FrontendOptions.td as well, here? Or Is it implied by being added to Options.td with the FrontendOption flag? I'm afraid I still don't really understand how the definitions in these two files, and the flags describing frontend-ness, relate.

(Will add HelpHidden, of course, good catch.)

jrose-apple replied:
The FrontendOption flag makes it available to the frontend when in Options.td; it's a default flag in FrontendOptions.td. (You can see how that works up at the top of FrontendOptions.td.) I just wasn't sure whether we needed or wanted a batch mode flag for the frontend at this point, or whether David had already added one (in which case I *think* the option-parsing system would complain that there were two flags with the same name). If this is the correct thing then go for it.

jrose-apple commented on include/swift/Driver/Compilation.h:
```diff
@@ -90,6 +90,11 @@ class Compilation {
+  /// The ToolChain, which the Compilation takes ownership of upon
+  /// construction, and which it may use to build subsequent batch
+  /// jobs itself.
+  std::unique_ptr<const ToolChain> TheToolChain;
```
This seems wrong to me. One Driver should (in theory) be able to perform multiple Compilations with the same ToolChain. I think that's more important than having Compilation stand on its own without a Driver.

(Also, I don't think `ToolChain` makes a const/non-const distinction to begin with, so it would be better to omit `const` here.)

graydon replied:
@jrose-apple Do you mean, by this, that I should copy the toolchain into the Compilation, by value? Or make it a shared pointer?

To be clear about the motive here: the issue isn't that I want there to be any particular numerical relationship between the number of ToolChains and Compilations in a given run; the issue is just that the ToolChain doesn't currently escape Driver::buildCompilation -- Compilation outlives ToolChain altogether -- and I need to be able to recover it (or "some" ToolChain) in order to re-enter Toolchain::constructInvocation and its callees, when building up a BatchJob, from within Compilation::performJobsImpl's callees.

I suppose I could also just re-manufacture a ToolChain from Driver anytime I need one, it .. looks like all the pieces are lying around for that to work. Would that be better?

jrose-apple replied:
I assumed the Driver owned both the ToolChain and the Compilation (well, by virtue of owning the stack), and therefore making ToolChain a reference here would be fine. A shared pointer would also be fine with me.

The original model for Compilation vs. Driver was that you could build a Compilation and execute it later, not that anyone would do that. (The thing you might do is build a Compilation, ask questions about it, and throw it away.) That's a little muddied now that the Compilation isn't "fully built" ahead of time.

None of this *really* matters. It's just object graph mental models.

jrose-apple replied:
(P.S. You can't copy the ToolChain by value because it's a virtual class hierarchy.)

graydon replied:
Ok, so you're suggesting I move the lifetime of Toolchain up the stack, to encompass the lifetime of Compilation, and then have Compilation hold a reference? That's fine by me (and I'll proceed that way until I hear otherwise here, trying not to waste your time too much!)

jrose-apple commented on include/swift/Driver/Compilation.h:
```diff
@@ -95,6 +96,11 @@ class Compilation {
+  /// The OutputInfo, which the Compilation takes ownership of upon
+  /// construction, and which it may use to build subsequent batch
+  /// jobs itself.
+  std::unique_ptr<OutputInfo> TheOutputInfo;
```
Ditto, ish. The Compilation isn't going to use this for anything *except* batching, and OutputInfo is a value-ish type anyway.

graydon replied:
@jrose-apple Same motive here; and here it would be more complex to re-manufacture the OutputInfo when needed later. Copy by-value?

jrose-apple reviewed (commented):
Hm. I don't like this from an object ownership perspective, but maybe you'll convince me otherwise.

jrose-apple commented:
Also I think it's misleading to put "NFC" in the PR title. Yes, this doesn't change the behavior of anything today, but it's also adding new stuff, not just moving existing stuff around.

(I'm making a distinction between the title and the description because the title is what people use to decide whether to click on it.)

graydon commented:
Re: "NFC", ok, changed title.

graydon committed:
[BatchMode] Hoist two args out of an inappropriate loop and switch-case.

graydon committed:
[BatchMode] Expand lifetime of ToolChain, pass reference to it to Compilation.

jrose-apple reviewed (commented):
Hm. Yeah, okay, I guess this makes sense.

graydon commented on lib/Driver/Driver.cpp:
```diff
@@ -501,16 +505,13 @@ static void validateEmbedBitcode(DerivedArgList &Args, OutputInfo &OI,
+                         std::unique_ptr<llvm::opt::InputArgList> ArgList) {
   llvm::PrettyStackTraceString CrashInfo("Compilation construction");
 
   llvm::sys::TimePoint<> StartTime = std::chrono::system_clock::now();
```
Yeah, it's missing the time taken to construct a single toolchain object though, which just initializes a couple pointers; I'm not worried about it.

graydon commented:
Rebased and fixed a test failure due to being imprecise about when to use primary file lists; now only doing so when necessary and in batch mode. I think everything's addressed in this set now.

graydon commented:
@swift-ci please test

jrose-apple commented on include/swift/Driver/Compilation.h:
```diff
@@ -70,6 +70,10 @@ class Compilation {
 
+  /// The ToolChain this Compilation was built with, that it may reuse to build
+  /// subsequent BatchJobs.
+  const ToolChain &TheToolChain;
```
Okay, now we're getting an unused variable warning for this. Can you add `LLVM_ATTRIBUTE_UNUSED` to this until the rest of it lands?
